Soft deactivation: Add cron job for running the deactivation management command.

adnrs96 opened the pull request.

timabbott commented on puppet/zulip/files/cron.d/soft-deactivate-users:
```diff
@@ -0,0 +1,6 @@
+USER=zulip
+
+# Cron job to soft deactivate long term idle users weekly.
+0 5 * * 7 zulip /home/zulip/deployments/current/manage.py soft_activate_deactivate_users -d --inactive-for 30
```
You also need to add this file to the `puppet` manifests, or it won't actually run.  I think you want it in `puppet/zulip/manifests/app_frontend.pp`.  

Also, to make sure this code doesn't run on staging servers, we should add a check for `settings.STAGING` in the management command, and have it do nothing if that's `True`.

timabbott replied:
Also, add `>/dev/null` at the end of the line, to suppress standard output.

timabbott commented on zerver/management/commands/soft_activate_deactivate_users.py:
```diff
@@ -63,18 +65,23 @@ def handle(self, *args, **options):
+                    users_to_deactivate = [
+                        user for user in users_to_deactivate
+                        if user.realm == realm
+                    ]
```
This algorithm will be pretty slow if we ask for a single realm on a large server, since we're filtering in Python, rather in the database.  

What if we instead pass into `get_users_for_soft_deactivation` a `filter_kwargs` argument, that's either `{}` or `dict(user_profile__realm=realm)`, and then add `**filter_kwargs` to the arguments in the `UserActivity.objects.filter()` line in get_users_for_soft_deactivation.

adnrs96 commented:
Thanks for reviewing this @timabbott .
I got to learn a new thing about building Django dynamic queries.
Addressed your comments here though I am not sure about `settings.STAGING`. I couldn't find it `settings.py` or elsewhere. How is this setting setup?

timabbott commented:
You need to rebase; I only added `settings.STAGING` recently.  It's in DEFAULT_SETTINGS.

timabbott commented on puppet/zulip/files/cron.d/soft-deactivate-users:
```diff
@@ -0,0 +1,6 @@
+USER=zulip
+
+# Cron job to soft deactivate long term idle users weekly.
+0 5 * * 7 zulip /home/zulip/deployments/current/manage.py soft_activate_deactivate_users -d --inactive-for 30 > /dev/null
```
traditionally, one usually writes it `>/dev/null`, not `> /dev/null`

timabbott replied:
Also, do we need the `--inactive-for` argument here?  Seems like we could just have it use the default.

adnrs96 replied:
We don't need it necessarily, I chose 28 days as default value for this reason only but then thought it might be worth specifying the argument there so that its more discoverable. I am just going to remove it now though.
Thanks for the `>/dev/null` advice. Though I am curious about the reason for doing so.

timabbott commented on zerver/management/commands/soft_activate_deactivate_users.py:
```diff
@@ -42,18 +42,27 @@ def add_arguments(self, parser):
-        realm = self.get_realm(options)
+        if settings.STAGING:
+            print('This is a Staging server. Suppressing management command.')
+            sys.exit(0)
```
I'd do a blank line after this, for readability.  

Probably a few other places in this function would benefit from similar treatment, separating section of the code.

adnrs96 commented:
@timabbott this has been updated.

timabbott commented:
Merged, thanks @adnrs96!
